lib/spack/spack/installer.py

```diff
@@ -1773,18 +1842,42 @@ def _next_is_pri0(self) -> bool:
         task = self.build_pq[0][1]
         return task.priority == 0
 
-    def _pop_task(self) -> Optional[Task]:
+    def _clear_removed_tasks(self):
+        """Get rid of any tasks in the queue with status 'BuildStatus.REMOVED'"""
+        while self.build_pq and self.build_pq[0][1].status == BuildStatus.REMOVED:
+            heapq.heappop(self.build_pq)
+
+    def _peek_ready_task(self) -> Optional[Task]:
         """
-        Remove and return the lowest priority task.
+        Return the first ready task in the queue, or None if there are no ready tasks.
+        """
+        self._clear_removed_tasks()
+        if not self.build_pq:
+            return None
 
-        Source: Variant of function at docs.python.org/2/library/heapq.html
+        task = self.build_pq[0][1]
+        return task if task.priority == 0 else None
+
+    def _pop_task(self) -> Task:
+        """Pop the first task off the queue and return it.
+
+        Raise an index error if the queue is empty."""
+        if not self.build_pq:
+            raise IndexError("Attempt to pop empty queue")
+        self._clear_removed_tasks()
```

Comment:
I *think* `_clear_removed_tasks()` should go first, as you do not know until after cleaning up removed tasks whether the queue if empty:

```suggestion
        self._clear_removed_tasks()
        if not self.build_pq:
            raise IndexError("Attempt to pop empty queue")
```